python/ray/dashboard/modules/actor/actor_head.py

```diff
@@ -310,21 +308,13 @@ def _convert_to_dict():
     async def _cleanup_actors(self):
         while True:
             try:
-                if len(DataSource.actors) > MAX_ACTORS_TO_CACHE:
-                    logger.debug("Cleaning up dead actors from GCS")
-                    while len(DataSource.actors) > MAX_ACTORS_TO_CACHE:
-                        if not self.dead_actors_queue:
-                            logger.warning(
-                                f"More than {MAX_ACTORS_TO_CACHE} "
-                                "live actors are cached"
-                            )
-                            break
-                        actor_id = self.dead_actors_queue.popleft()
-                        if actor_id in DataSource.actors:
-                            actor = DataSource.actors.pop(actor_id)
-                            node_id = actor["address"].get("rayletId")
-                            if node_id and node_id != actor_consts.NIL_NODE_ID:
-                                del DataSource.node_actors[node_id][actor_id]
+                while len(self.dead_actors_queue) > MAX_DELETED_ACTORS_TO_CACHE:
```

Comment:
Is it possible that GCS and dashboard evict different dead actors?

Reply:
ya good point, so post python 3.7, dictionaries preserve insertion order, and they're coming in in-order from the gcs actor channel in the same order that they're added to the gcs's dead actor queue because the gcs both publishes and adds to cache inside the actor manager's DestroyActor function, so both queues should have the same order, and then both are just popping from the front whenever it goes over the max. 

One concern I have is that gcs is only adding to it's dead actor cache if the actor is not restartable, but the publish happens either way, can non-restartable actors still be marked dead? If so they'll end up with different dead actor queues.

This does feel like a very weak guarantee though. And I don't think it makes sense to expect people to keep the dashboard queue in mind while making changes in the gcs queue. I think two possible solutions for a stronger relationship are having a minheap based on death timestamp (has to be new pushed field), or having the dashboard not keep it's own cache. But I see alexey recently revamped this just last month for performance reasons.

https://github.com/ray-project/ray/pull/47617/files#diff-0df41fb197ae91ea90ce4414703c9bb809d01ac60f01639fe0bdaca4e4331b1e

Reply:
Let's defer this to the next PR since we have the same issue today.